pkg/netpol/eval/internal/k8s/representative_selectors.go

```diff
@@ -0,0 +1,120 @@
+// note that : for non-empty selectors that partly match/ contained in each other the func returns false
+func SelectorsFullMatch(ruleSelector, repSelector *v1.LabelSelector) (bool, error) {
+	if ruleSelector == repSelector { // both label selectors point to same reference
+		return true, nil
+	}
+	// otherwise, check requirements equality
+	selector1, err := v1.LabelSelectorAsSelector(ruleSelector)
+	if err != nil {
+		return false, err
+	}
+	if selector1.Empty() { // empty rule matches everything
+		return true, nil
+	}
+	selector2, err := v1.LabelSelectorAsSelector(repSelector)
+	if err != nil {
+		return false, err
+	}
+	// Requirements() returns sorted by key list;
+	// since `v1.LabelSelectorAsSelector` builds a new `labels.selector` by using `Requirements.Add` which sorts the requirements slice by key.
+	// and `Requirements()` returns this sorted list
+	// links:
+	// v1.LabelSelectorAsSelector:
+	// https://github.com/kubernetes/apimachinery/blob/dc7e034c86479d49be4b0eefad307621e10caa0e/pkg/apis/meta/v1/helpers.go#L34
+	// Requirements.Add : https://github.com/kubernetes/apimachinery/blob/d7e1c5311169d5ece2db0ae0118066859aa6f7d8/pkg/labels/selector.go#L373
+	requirements1, _ := selector1.Requirements() // sorted
+	requirements2, _ := selector2.Requirements() // sorted
+	if len(requirements1) != len(requirements2) {
+		return false, nil
+	}
+	for i := range requirements1 {
+		// requirements1[i].Equal(requirements2[i]) returns false if the values are not in same order
+		// (stringslices.Equal returns true only if two slices have same length and same order of items)
+		// however, `Requirement.String` sorts the values using `safeSort`, so will compare by string
+		// link to Requirement.String() :
+		// https://github.com/kubernetes/apimachinery/blob/d7e1c5311169d5ece2db0ae0118066859aa6f7d8/pkg/labels/selector.go#L310
+		str1 := requirements1[i].String()
+		str2 := requirements2[i].String()
+
+		// handling special case of one requirement has `In` operator with 1 value, and other requirement has "=" operator (from matchLabel)
+		// for example : req1 := (app in x), req2 := (app=x) >> we expect a full match, however,
+		// labels pkg treats them as different requirements
+		// i.e. requirements1[i].Equal(requirements2[i]) and requirements1[i].String() != requirements2[i].String() return false in this case
+		// requirement.String() : returns <key>=<values> string for "Equals" operator and returns (<key> in (<values));
+		// so, in case on requirement is "in" with one value only , will convert its string to the <key>=<values> format to get correct result
+		if newStr1 := handleRequirementWithInOpAndSingleValue(requirements1[i]); newStr1 != "" {
+			str1 = newStr1
+		}
+		if newStr2 := handleRequirementWithInOpAndSingleValue(requirements2[i]); newStr2 != "" {
+			str2 = newStr2
+		}
+		if str1 != str2 {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
```

Comment:
where does this str come from? better build a relevant selector object and use its string() function.

Reply:
as described in the comments `string()` func of selector is not helpful for our case.
we loop sorted selector's requirements and compare strings of the requirements. (if not all requirements strings equal, then selectors not equal).

the use of `handleRequirementWithInOpAndSingleValue` is only for a single case of a single requirement. what it does actually:
if the requirement applies the condition (In operator with single value) it replaces its string to be <key> = <val> instead of <key> in <val>.
so we ensure two requirements {<key> in <val>} and {<key> = <val>} are treated the same.

renamed the func to `replaceStringOfRequirementWithInOpAndSingleValue`

is this more clear?

(creating a new selector with all other requirements too may be time costing; and we can't predict if any of the requirements will have{<key> in <val>} )

Reply:
please generate a new req object , to get the string from it

Reply:
done